content/academy/data_collection_basics/api_scraping.md

```diff
@@ -0,0 +1,223 @@
+
+```JavaScript
+const response = await gotScraping({
+    url: 'https://example.com',
+    headerGeneratorOptions:{
+        browsers: [
+            {
+                name: 'chrome',
+                minVersion: 87,
+                maxVersion: 89
+            }
+        ],
+        devices: ['desktop'],
+        locales: ['de-DE', 'en-US'],
+        operatingSystems: ['windows', 'linux'],
+    },
+    headers: {
+        'some-header': 'Hello, Academy!'
+    }
+})
+```
+
+For our SoundCloud example, testing the endpoint from the previous section in a tool like [Postman](https://www.postman.com/) works perfectly, and returns the data we want; however, when the "`client_id`" parameter is removed, we receive a `401 Unauthorized` error. Luckily, the Client ID is the same for every user, which means that it is not tied to a session or an IP address (this is based on our own observations and tests). The big downfall is that the token being used by SoundCloud changes every few weeks, so it shouldn't be hardcoded. This case is actually quite common, and is not only seen with SoundCloud.
+
+Ideally, this `client_id` should be scraped dynamically, especially since it changes frequently, but unfortunately, the token cannot be found anywhere on SoundCloud's pages. We already know that it's available within the parameters of certain requests though, and luckily, [Puppeteer](https://github.com/puppeteer/puppeteer) offers a simple way to analyze each response when on a page. It's a bit like using browser DevTools, which you are already familiar with by now, but programatically instead.
+
+Here is a way you could dynamically scrape the `client_id` using Puppeteer:
+
+```JavaScript
+const puppeteer = require('puppeteer'); // import the puppeteer module
+
+const scrapeClientId = async () => {
+    const browser = await puppeteer.launch({ headless: false });
+    const page = await browser.newPage();
+
+    let clientId = null; // initialize a variable that will eventually hold the client_id
+
+    page.on('response', async (res) => { // handle each response
+        const id = new URL(res._url).searchParams.get('client_id') ?? null; // try to grab the 'client_id' parameter from each URL
+
+        if (id) clientId = id; // if the parameter exists, set our clientId variable to the newly parsed value
+    });
+
+    await page.goto('https://soundcloud.com/tiesto/tracks'); // visit the page
+    await page.waitForSelector('.sc-classic'); // wait for a selector that ensures the page has time to load and make requests to its API
+
+    await browser.close();
+    console.log(clientId); // log the retrieved client_id
+};
+
+scrapeClientId();
+```
+
+# [](#extra-challenges) Extra challenges
+
+> **Note:** these concepts will be covered more in-depth later on, but it is good to be aware of them now.
```

Comment:
It's generally good practice to add some text between two headings, so maybe this doesn't have to be a note? Instead, maybe you could introduce what's to come and add this at the end?